File: frontend/src/api/summary.ts
import apiClient from './client';

// --- Radar Scores ---
export interface RadarScores {
  fluency: number;
  vocabulary: number;
  grammar: number;
  pronunciation: number;
  interaction: number;
}

// --- Highlight ---
export interface Highlight {
  title: string;
  description: string;
  example_sentence?: string;
}

// --- Practice Suggestion (V1.1) ---
export interface PracticeSuggestion {
  type: string;
  target: string;
  suggested_exercise_id: string | null;
}

// --- Pronunciation Error ---
export interface PronunciationErrorItem {
  utterance_id: string;
  sentence: string;
  score: number;
  detail_url: string;
}

// --- Grammar Error ---
export interface GrammarErrorItem {
  utterance_id: string;
  original: string;
  error_type: string;
  error_span: Record<string, number>;
  correction: string;
  corrected_sentence: string | null;
  explanation: string | null;
  severity: string;
}

// --- Coach Card (light coach loop) ---
export interface CoachCardInsight {
  type: 'strength' | 'improvement';
  text: string;
  detail?: string | null;
}

export interface CoachCard {
  scene_name: string;
  duration_seconds: number;
  overall_score: number | null;
  strengths: CoachCardInsight[];
  improvements: CoachCardInsight[];
  score_delta: number | null;
  chat_only?: boolean;
}

export interface NextPracticeOption {
  kind: 'consolidate' | 'explore';
  scene_id: number;
  scene_name: string;
  label: string;
  reason: string;
}

export interface NextPracticeResponse {
  options: NextPracticeOption[];
}

// --- Session Summary (V1.1: radar_scores) ---
export interface SessionSummary {
  id: string;
  session_id: string;
  scene_name: string | null;
  duration_seconds: number;
  radar_scores: RadarScores;
  highlights: Highlight[];
  top_pronunciation_errors: PronunciationErrorItem[];
  top_grammar_errors: GrammarErrorItem[];
  practice_suggestions: PracticeSuggestion[];
  share_image_url: string | null;
  coach_card?: CoachCard | null;
  next_practice_options?: NextPracticeOption[];
  created_at: string;
}

// --- Progress Snapshot ---
export interface ProgressSnapshot {
  snapshot_date: string;
  total_score: number;
  dimension_scores: Record<string, number>;
  session_count: number;
  total_duration_seconds: number;
}

// --- Weakness Record ---
export interface WeaknessRecord {
  period_start: string;
  period_end: string;
  category: string;
  item: string;
  error_count: number;
  trend: string | null;
}

// --- User Progress ---
export interface ProgressProvenance {
  pronunciation: 'iflytek_ise' | 'text_analysis' | 'none';
  fluency: 'iflytek_ise' | 'text_analysis' | 'none';
  grammar: 'evaluation' | 'none';
  vocabulary: 'heuristic' | 'none';
  interaction: 'heuristic' | 'none';
  strengths: 'computed' | 'none';
}

export interface DimensionAvailable {
  pronunciation: boolean;
  fluency: boolean;
  grammar: boolean;
  vocabulary: boolean;
  interaction: boolean;
}

export interface UserProgress {
  user_id: string;
  overall_rating: string;
  total_score: number;
  total_sessions: number;
  total_hours: number;
  snapshots: ProgressSnapshot[];
  weaknesses: WeaknessRecord[];
  strengths: Record<string, unknown>[];
  data_provenance?: ProgressProvenance;
  dimension_available?: DimensionAvailable;
}

// --- Achievement (V1.1) ---
export interface AchievementInfo {
  id: string;
  title: string;
  description: string;
  current_progress: number;
  target: number;
  unlocked_at: string | null;
  icon_url: string | null;
  icon?: string | null;
}

export interface AchievementList {
  user_id: string;
  achievements: AchievementInfo[];
}

// --- Trend ---
export interface TrendPoint {
  date: string;
  score: number;
  dimension: string | null;
}

export interface ForecastPoint {
  date: string;
  score: number;
  confidence_interval_lower: number;
  confidence_interval_upper: number;
}

export interface ProgressTrend {
  dimension: string;
  granularity: string;
  data_points: TrendPoint[];
  forecast: ForecastPoint[];
}

export interface TrendParams {
  start_date: string;
  end_date: string;
  dimension?: string;
  granularity?: string;
  include_forecast?: boolean;
}

// --- Weakness Distribution (V1.1) ---
export interface WeaknessDistItem {
  category: string;
  item: string;
  error_count: number;
  trend: string | null;
  suggested_exercise_id: string | null;
}

export interface WeaknessDistResponse {
  user_id: string;
  period: string;
  weakness_matrix: WeaknessDistItem[];
}

export interface WeaknessDistParams {
  start_date: string;
  end_date: string;
  category?: string;
}

// --- Review Plan ---
export interface ReviewPlanItem {
  type: string;
  target: string;
  exercise_ids: string[];
  estimated_minutes: number;
}

export interface ReviewPlan {
  plan_id: string;
  generated_at: string;
  items: ReviewPlanItem[];
}

export const summaryApi = {
  getSessionSummary(sessionId: string) {
    return apiClient
      .get<SessionSummary>(`/sessions/${sessionId}/summary`)
      .then((res) => res.data);
  },

  getNextPractice(userId: string) {
    return apiClient
      .get<NextPracticeResponse>(`/users/${userId}/next-practice`)
      .then((res) => res.data);
  },

  getProgress(userId: string) {
    return apiClient
      .get<UserProgress>(`/users/${userId}/progress`)
      .then((res) => res.data);
  },

  getAchievements(userId: string) {
    return apiClient
      .get<AchievementList>(`/users/${userId}/achievements`)
      .then((res) => res.data);
  },

  getProgressTrend(userId: string, params: TrendParams) {
    return apiClient
      .get<ProgressTrend>(`/users/${userId}/progress/trend`, { params })
      .then((res) => res.data);
  },

  getWeaknessDistribution(userId: string, params: WeaknessDistParams) {
    return apiClient
      .get<WeaknessDistResponse>(`/users/${userId}/progress/weakness-distribution`, { params })
      .then((res) => res.data);
  },

  /** V1.1: Get personalized review plan */
  getReviewPlan(userId: string) {
    return apiClient
      .get<ReviewPlan>(`/users/${userId}/review-plan`)
      .then((res) => res.data);
  },
};

File: backend/app/schemas/summary.py
import uuid
from datetime import date, datetime
from typing import Optional

from typing import Literal

from pydantic import BaseModel, Field


# --- Session Summary ---

class RadarScores(BaseModel):
    fluency: int = Field(ge=0, le=100)
    vocabulary: int = Field(ge=0, le=100)
    grammar: int = Field(ge=0, le=100)
    pronunciation: int = Field(ge=0, le=100)
    interaction: int = Field(ge=0, le=100)


class Highlight(BaseModel):
    title: str
    description: str
    example_sentence: str | None = None


class PracticeSuggestion(BaseModel):
    type: str
    target: str
    suggested_exercise_id: str | None = None


class TopPronunciationError(BaseModel):
    utterance_id: uuid.UUID
    sentence: str
    score: int
    detail_url: str


class TopGrammarError(BaseModel):
    utterance_id: uuid.UUID
    original: str
    error_type: str
    error_span: dict
    correction: str
    corrected_sentence: str | None = None
    explanation: str | None = None
    severity: str = "medium"


class CoachCardInsight(BaseModel):
    type: Literal["strength", "improvement"]
    text: str
    detail: str | None = None


class CoachCardResponse(BaseModel):
    scene_name: str
    duration_seconds: int
    overall_score: int | None = Field(default=None, ge=0, le=100)
    strengths: list[CoachCardInsight] = []
    improvements: list[CoachCardInsight] = []
    score_delta: int | None = None
    chat_only: bool = False


class NextPracticeOption(BaseModel):
    kind: Literal["consolidate", "explore"]
    scene_id: int
    scene_name: str
    label: str
    reason: str


class NextPracticeResponse(BaseModel):
    options: list[NextPracticeOption]


class SessionSummaryResponse(BaseModel):
    id: uuid.UUID
    session_id: uuid.UUID
    scene_name: str | None = None
    duration_seconds: int = 0
    radar_scores: RadarScores
    highlights: list[Highlight] = []
    top_pronunciation_errors: list[TopPronunciationError] = []
    top_grammar_errors: list[TopGrammarError] = []
    practice_suggestions: list[PracticeSuggestion] = []
    share_image_url: str | None = None
    coach_card: CoachCardResponse | None = None
    next_practice_options: list[NextPracticeOption] = []
    created_at: datetime

    class Config:
        from_attributes = True


# --- User Progress ---

class ProgressSnapshot(BaseModel):
    snapshot_date: date
    total_score: int
    dimension_scores: dict  # {"fluency": 60, "vocabulary": 55, ...}
    session_count: int
    total_duration_seconds: int


class WeaknessRecord(BaseModel):
    period_start: date
    period_end: date
    category: str  # "pronunciation" or "grammar"
    item: str
    error_count: int
    trend: str | None = None  # "rising", "falling", "stable"


class ProgressProvenance(BaseModel):
    pronunciation: str = "none"  # iflytek_ise | text_analysis | none
    fluency: str = "none"
    grammar: str = "none"  # evaluation | none
    vocabulary: str = "none"  # heuristic | none
    interaction: str = "none"  # heuristic | none
    strengths: str = "none"  # computed | none


class DimensionAvailable(BaseModel):
    pronunciation: bool = False
    fluency: bool = False
    grammar: bool = False
    vocabulary: bool = False
    interaction: bool = False


class UserProgressResponse(BaseModel):
    user_id: uuid.UUID
    overall_rating: str  # "A1"-"C2" or "beginner"-"advanced"
    total_score: int = Field(ge=0, le=100)
    total_sessions: int
    total_hours: float
    snapshots: list[ProgressSnapshot] = []
    weaknesses: list[WeaknessRecord] = []
    strengths: list[dict] = []
    data_provenance: ProgressProvenance = Field(default_factory=ProgressProvenance)
    dimension_available: DimensionAvailable = Field(default_factory=DimensionAvailable)


# --- Achievements ---

class AchievementInfo(BaseModel):
    id: str
    title: str
    description: str
    current_progress: int = 0
    target: int = 100
    unlocked_at: datetime | None = None
    icon_url: str | None = None
    icon: str | None = None

    class Config:
        from_attributes = True


class AchievementListResponse(BaseModel):
    user_id: uuid.UUID
    achievements: list[AchievementInfo] = []


# --- Progress Trend ---

class TrendPoint(BaseModel):
    date: date
    score: int
    dimension: str | None = None  # None when dimension="all"


class ForecastPoint(BaseModel):
    date: date
    score: int
    confidence_interval_lower: int
    confidence_interval_upper: int


class ProgressTrendResponse(BaseModel):
    dimension: str = "total_score"
    granularity: str = "daily"
    data_points: list[TrendPoint] = []
    forecast: list[ForecastPoint] = []


# --- Weakness Distribution ---

class WeaknessDistItem(BaseModel):
    category: str
    item: str
    error_count: int
    trend: str | None = None
    suggested_exercise_id: str | None = None


class WeaknessDistResponse(BaseModel):
    user_id: uuid.UUID
    period: str = "last_month"
    weakness_matrix: list[WeaknessDistItem] = []
